drivers/ethernet/eth_stm32_hal_v2.c

```diff
@@ -0,0 +1,637 @@
+static struct eth_stm32_rx_buffer_header dma_rx_buffer_header[ETH_RXBUFNB];
+static struct eth_stm32_tx_buffer_header dma_tx_buffer_header[ETH_TXBUFNB];
+static struct eth_stm32_tx_context dma_tx_context[ETH_TX_DESC_CNT];
+
+void HAL_ETH_RxAllocateCallback(uint8_t **buf)
+{
+	for (size_t i = 0; i < ETH_RXBUFNB; ++i) {
+		if (!dma_rx_buffer_header[i].used) {
+			dma_rx_buffer_header[i].next = NULL;
+			dma_rx_buffer_header[i].size = 0;
+			dma_rx_buffer_header[i].used = true;
+			*buf = dma_rx_buffer[i];
+			return;
+		}
+	}
+	*buf = NULL;
+}
+
+/* Pointer to an array of ETH_STM32_RX_BUF_SIZE uint8_t's */
+typedef uint8_t (*RxBufferPtr)[ETH_STM32_RX_BUF_SIZE];
+
+/* called by HAL_ETH_ReadData() */
+void HAL_ETH_RxLinkCallback(void **pStart, void **pEnd, uint8_t *buff, uint16_t Length)
+{
+	/* buff points to the begin on one of the rx buffers,
+	 * so we can compute the index of the given buffer
+	 */
+	size_t index = (RxBufferPtr)buff - &dma_rx_buffer[0];
+	struct eth_stm32_rx_buffer_header *header = &dma_rx_buffer_header[index];
+
+	__ASSERT_NO_MSG(index < ETH_RXBUFNB);
+
+	header->size = Length;
+
+	if (!*pStart) {
+		/* first packet, set head pointer of linked list */
+		*pStart = header;
+		*pEnd = header;
+	} else {
+		__ASSERT_NO_MSG(*pEnd != NULL);
+		/* not the first packet, add to list and adjust tail pointer */
+		((struct eth_stm32_rx_buffer_header *)*pEnd)->next = header;
+		*pEnd = header;
+	}
+}
+
+/* Called by HAL_ETH_ReleaseTxPacket */
+void HAL_ETH_TxFreeCallback(uint32_t *buff)
+{
+	__ASSERT_NO_MSG(buff != NULL);
+
+	/* buff is the user context in tx_config.pData */
+	struct eth_stm32_tx_context *ctx = (struct eth_stm32_tx_context *)buff;
+	struct eth_stm32_tx_buffer_header *buffer_header =
+		&dma_tx_buffer_header[ctx->first_tx_buffer_index];
+
+	while (buffer_header != NULL) {
+		buffer_header->used = false;
+		if (buffer_header->tx_buff.next != NULL) {
+			buffer_header = CONTAINER_OF(buffer_header->tx_buff.next,
+				struct eth_stm32_tx_buffer_header, tx_buff);
+		} else {
+			buffer_header = NULL;
+		}
+	}
+	ctx->used = false;
+}
+
+/* allocate a tx buffer and mark it as used */
+static inline uint16_t allocate_tx_buffer(void)
+{
+	for (;;) {
+		for (uint16_t index = 0; index < ETH_TXBUFNB; index++) {
+			if (!dma_tx_buffer_header[index].used) {
+				dma_tx_buffer_header[index].used = true;
+				return index;
+			}
+		}
+		k_yield();
+	}
+}
+
+/* allocate a tx context and mark it as used, the first tx buffer is also allocated */
+static inline struct eth_stm32_tx_context *allocate_tx_context(struct net_pkt *pkt)
+{
+	for (;;) {
+		for (uint16_t index = 0; index < ETH_TX_DESC_CNT; index++) {
+			if (!dma_tx_context[index].used) {
+				dma_tx_context[index].used = true;
+				dma_tx_context[index].pkt = pkt;
+				dma_tx_context[index].first_tx_buffer_index = allocate_tx_buffer();
+				return &dma_tx_context[index];
+			}
+		}
+		k_yield();
+	}
+}
+
+#if defined(CONFIG_ETH_STM32_HAL_API_V2)
+static ETH_TxPacketConfig tx_config;
```

Comment:
This is used only far below. Move to source file top, next to line 46?

Reply:
Done